### PR TITLE
Add `[GeneratePublicApi]` to `TracerSettings`

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

- Add `PublicApiUsages` enum values
- Update `TracerSettings` et al. to create "mirrored" internal properties, so we can track public api usage
- Make `internal` properties `get` only where possible.
- Added a helper function `TracerSettings.Create()` for more easily creating a `TracerSettings` object (primarily for use in tests)

## Reason for change

We want to know which APIs customers are using so we can more safely deprecate/evolve the public API surface. WE do that using telemetry metrics. 

Unfortunately, that means that _we_ mustn't call the APIs, otherwise we will skew our metrics, which means we need to have "internal" versions of our public APIs that we use. 

Additionally, made as many of the `internal` properties of `TracerSettings`, `ExporterSettings`, and `IntegrationSettings` be `get` only. This is to ensure that _we_ don't change them in code (unless absolutely necessary) because if a property is "changeable" we need to track the new value in telemetry config (subsequent PR). 

To reduce some of the pain making things `get`-only causes on the testing side, introduced a `TracerSettings.Create()` helper method that makes it easier to pass in key-value config.

## Implementation details

Used the `[GeneratePublicApi]` source generator attribute to create "wrapper" properties (which mimic the current public API) and delegate access to the "internal" properties.

In some cases (where we don't have auto-properties) we can't use `[GeneratePublicApi]`, so we instrument the property directly.

Updated tests where necessary to use `TracerSettings.Create()` instead of setting now-immutable properties.

## Test coverage

Primarily covered by existing tests. We can still use the public APIs in tests so nothing fundamentally really changes with that.

## Other details

This doesn't attempt to record all public API usages yet - that will come in a separate PR. This is primarily focused on prerequisites for config telemetry.

Stacked on https://github.com/DataDog/dd-trace-dotnet/pull/4214

Supersedes https://github.com/DataDog/dd-trace-dotnet/pull/4215

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (4219)](https://github.com/DataDog/dd-trace-dotnet/tree/621dfaa16f9f03eb9b8f1ef91ab889c7cf195c10)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/e1fea541a6ef8e301b7016281f500eef049eba03)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (3,038ms)  : 2965, 3110
     .   : milestone, 3038,
    master - mean (3,009ms)  : 2907, 3111
     .   : milestone, 3009,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (3,788ms)  : 3665, 3910
     .   : milestone, 3788,
    master - mean (3,808ms)  : 3742, 3874
     .   : milestone, 3808,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (3,134ms)  : 2999, 3269
     .   : milestone, 3134,
    master - mean (3,165ms)  : 3028, 3303
     .   : milestone, 3165,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (3,599ms)  : 3540, 3658
     .   : milestone, 3599,
    master - mean (3,616ms)  : 3528, 3704
     .   : milestone, 3616,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (3,114ms)  : 2996, 3231
     .   : milestone, 3114,
    master - mean (3,098ms)  : 2974, 3222
     .   : milestone, 3098,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (3,578ms)  : 3488, 3668
     .   : milestone, 3578,
    master - mean (3,565ms)  : 3498, 3633
     .   : milestone, 3565,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (189ms)  : 184, 193
     .   : milestone, 189,
    master - mean (187ms)  : 184, 190
     .   : milestone, 187,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (1,032ms)  : 959, 1106
     .   : milestone, 1032,
    master - mean (1,037ms)  : 961, 1113
     .   : milestone, 1037,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (367ms)  : 362, 372
     .   : milestone, 367,
    master - mean (367ms)  : 362, 373
     .   : milestone, 367,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (1,139ms)  : 1109, 1169
     .   : milestone, 1139,
    master - mean (1,139ms)  : 1115, 1163
     .   : milestone, 1139,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4219) - mean (357ms)  : 349, 366
     .   : milestone, 357,
    master - mean (355ms)  : 349, 360
     .   : milestone, 355,

    section CallTarget+Inlining+NGEN
    This PR (4219) - mean (1,092ms)  : 1041, 1143
     .   : milestone, 1092,
    master - mean (1,091ms)  : 1040, 1143
     .   : milestone, 1091,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4219)](https://github.com/DataDog/dd-trace-dotnet/tree/621dfaa16f9f03eb9b8f1ef91ab889c7cf195c10)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/e1fea541a6ef8e301b7016281f500eef049eba03)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)
- [benchmarks/2.31.0](https://github.com/DataDog/dd-trace-dotnet/tree/5a0f7f5a6023b3b0761baa6b1bcd53b8f3af5655)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4219) (11.134M)   : 0, 11133956
    master (11.047M)   : 0, 11047114
    benchmarks/2.31.0 (11.076M)   : 0, 11075765
    benchmarks/2.9.0 (11.213M)   : 0, 11213357

    section Automatic
    This PR (4219) (7.831M)   : 0, 7830853
    master (7.800M)   : 0, 7799621
    benchmarks/2.31.0 (7.710M)   : 0, 7709708
    benchmarks/2.9.0 (8.000M)   : 0, 7999700

    section Trace stats
    master (7.779M)   : 0, 7778936
    benchmarks/2.31.0 (7.800M)   : 0, 7800275

    section Manual
    This PR (4219) (9.901M)   : 0, 9900635
    master (9.837M)   : 0, 9836744
    benchmarks/2.31.0 (9.918M)   : 0, 9918073

    section Manual + Automatic
    This PR (4219) (7.494M)   : 0, 7494185
    master (7.490M)   : 0, 7490095
    benchmarks/2.31.0 (7.411M)   : 0, 7411448

    section Version Conflict
    master (6.775M)   : 0, 6774777
    benchmarks/2.31.0 (6.700M)   : 0, 6700156

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4219) (9.773M)   : 0, 9772548
    master (9.603M)   : 0, 9602840
    benchmarks/2.31.0 (9.734M)   : 0, 9734207
    benchmarks/2.9.0 (9.677M)   : 0, 9676916

    section Automatic
    This PR (4219) (6.857M)   : 0, 6856928
    master (6.781M)   : 0, 6781389
    benchmarks/2.31.0 (6.747M)   : 0, 6747040

    section Trace stats
    master (6.775M)   : 0, 6775417
    benchmarks/2.31.0 (6.884M)   : 0, 6883900

    section Manual
    This PR (4219) (8.404M)   : 0, 8404395
    master (8.491M)   : 0, 8490992
    benchmarks/2.31.0 (8.527M)   : 0, 8527220

    section Manual + Automatic
    This PR (4219) (6.558M)   : 0, 6557869
    master (6.729M)   : 0, 6728809
    benchmarks/2.31.0 (6.585M)   : 0, 6585247

    section Version Conflict
    master (5.924M)   : 0, 5923699
    benchmarks/2.31.0 (5.807M)   : 0, 5807273

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4219) (10.393M)   : 0, 10393304
    master (10.048M)   : 0, 10047810
    benchmarks/2.31.0 (10.179M)   : 0, 10178625
    benchmarks/2.9.0 (10.236M)   : 0, 10236148

    section Automatic
    This PR (4219) (7.630M)   : 0, 7630222
    master (7.543M)   : 0, 7542770
    benchmarks/2.31.0 (7.339M)   : 0, 7339035
    benchmarks/2.9.0 (7.503M)   : 0, 7503041

    section Trace stats
    master (7.555M)   : 0, 7555289
    benchmarks/2.31.0 (7.460M)   : 0, 7459503

    section Manual
    This PR (4219) (9.445M)   : 0, 9444752
    master (9.226M)   : 0, 9225550
    benchmarks/2.31.0 (9.546M)   : 0, 9545617

    section Manual + Automatic
    This PR (4219) (7.364M)   : 0, 7363531
    master (7.329M)   : 0, 7328912
    benchmarks/2.31.0 (7.410M)   : 0, 7409995

    section Version Conflict
    master (6.647M)   : 0, 6647066
    benchmarks/2.31.0 (6.644M)   : 0, 6643662

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4219) (7.574M)   : 0, 7573554
    master (7.480M)   : 0, 7479904
    benchmarks/2.31.0 (7.370M)   : 0, 7370014
    benchmarks/2.9.0 (7.848M)   : 0, 7848262

    section No attack
    This PR (4219) (2.396M)   : 0, 2395512
    master (2.405M)   : 0, 2404829
    benchmarks/2.31.0 (2.400M)   : 0, 2400385
    benchmarks/2.9.0 (3.145M)   : 0, 3144618

    section Attack
    This PR (4219) (2.029M)   : 0, 2028997
    master (2.053M)   : 0, 2053413
    benchmarks/2.31.0 (2.043M)   : 0, 2043215
    benchmarks/2.9.0 (2.531M)   : 0, 2531234

    section Blocking
    This PR (4219) (4.114M)   : 0, 4113725
    master (4.145M)   : 0, 4144782
    benchmarks/2.31.0 (4.150M)   : 0, 4149653

```

File: tracer/src/Datadog.Trace/Telemetry/Metrics/PublicApiUsage.cs [hunk 1]
```diff
@@ -16,9 +15,176 @@ namespace Datadog.Trace.Telemetry.Metrics;
...
     [Description("tracer_ctor")] Tracer_Ctor,
     [Description("tracer_ctor_settings")] Tracer_Ctor_Settings,
     [Description("tracer_instance_set")] Tracer_Instance_Set,
     [Description("tracer_configure")] Tracer_Configure,
-    // TODO: more will be added soon, these are just for testing initially
+    [Description("tracer_forceflushasync")] Tracer_ForceFlushAsync,
+    [Description("tracer_startactive")] Tracer_StartActive,
+    [Description("tracer_startactive_settings")] Tracer_StartActive_Settings,
+
+    // These are problematic, as we use them _everywhere_ so means a lot of code changes
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Telemetry/Metrics/PublicApiUsage.cs:29]: Yeah, I'm not actually _using_ all of these yet - I'm going to do another PR that attempts to add all these properly 🙂 The downside is that these are all ripe for conflicts 😂

File: tracer/src/Datadog.Trace/Configuration/ImmutableTracerSettings.cs [hunk 2]
```diff
@@ -43,48 +43,48 @@ public ImmutableTracerSettings(IConfigurationSource source)
...
+            HeaderTags = new ReadOnlyDictionary<string, string>(settings.HeaderTagsInternal);
+            GrpcTags = new ReadOnlyDictionary<string, string>(settings.GrpcTagsInternal);
             IpHeader = settings.IpHeader;
             IpHeaderEnabled = settings.IpHeaderEnabled;
-            TracerMetricsEnabled = settings.TracerMetricsEnabled;
-            StatsComputationEnabled = settings.StatsComputationEnabled;
+            TracerMetricsEnabled = settings.TracerMetricsEnabledInternal;
+            StatsComputationEnabled = settings.StatsComputationEnabledInternal;
             StatsComputationInterval = settings.StatsComputationInterval;
             RuntimeMetricsEnabled = settings.RuntimeMetricsEnabled;
```
[the remark this answers is not in the document]
Comment (pierotibou) [reply tracer/src/Datadog.Trace/Configuration/ImmutableTracerSettings.cs:85]: Completely agree on the cosmetic aspect. Just asking what seems the more natural for everyone. I mean once you know it, the current state works out well. I'm fine either way TBH

**Review by kevingosse — APPROVED** [review]
Thanks I hate it 😛

**Pull request closed without merging**